tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15797,16 +15857,20 @@ clang::FunctionDecl *ValidateNoRecursion(clang::Sema *self,
   return nullptr;
 }
 
+// There is an assumption being made that this function will only be called
+// on library shader stages.Thus, there is no entry point, only exported
+// functions
 void ValidateNoRecursionInTranslationUnit(clang::Sema *self) {
   std::set<FunctionDecl *> FDecls;
   std::vector<FunctionDecl *> FDeclsVec;
   for (auto decl : self->getASTContext().getTranslationUnitDecl()->decls()) {
-    // TODO: improve condition so that only exported functions are checked,
-    // instead of all functions. Issue: #5857
     if (FunctionDecl *FD = dyn_cast<FunctionDecl>(decl)) {
       // returns the first recursive function declaration detected
       // from this function declaration FD, and determines whether
       // the recursion was detected in the patch-constant function
+      if (!IsExported(self, FD)) {
+        continue;
+      }
       FunctionDecl *pResult = ValidateNoRecursion(self, FD);
```

Comment:
We also need to add a map to `CallGraphWithRecurseGuard` to keep track of recursive functions we have already identified, since this was originally designed with only one entry point in mind.  We can use this map at the beginning of CheckRecursion to early out if we already checked this function before.

Something like this:
```diff
@@ -2916,8 +2916,9 @@ struct CallNode {
 typedef ::llvm::DenseMap<FunctionDecl *, CallNode> CallNodes;
 typedef ::llvm::SmallPtrSet<Decl *, 8> FnCallStack;
 typedef ::llvm::SmallPtrSet<FunctionDecl *, 128> FunctionSet;
 typedef ::llvm::SmallVector<FunctionDecl *, 32> PendingFunctions;
+typedef ::llvm::DenseMap<FunctionDecl *, FunctionDecl *> FunctionMap;
 
 // Returns the definition of a function.
 // This serves two purposes - ignore built-in functions, and pick
 // a single Decl * to be used in maps and sets.
@@ -2990,21 +2991,28 @@ public:
 class CallGraphWithRecurseGuard {
 private:
   CallNodes m_callNodes;
   FunctionSet m_visitedFunctions;
+  FunctionMap m_functionsCheckedForRecursion;
 
   FunctionDecl *CheckRecursion(FnCallStack &CallStack, FunctionDecl *D) const {
+    auto it = m_functionsCheckedForRecursion.find(D);
+    if (it != m_functionsCheckedForRecursion.end())
+      return it->second;
     if (CallStack.insert(D).second == false)
       return D;
     auto node = m_callNodes.find(D);
     if (node != m_callNodes.end()) {
       for (FunctionDecl *Callee : node->second.CalleeFns) {
         FunctionDecl *pResult = CheckRecursion(CallStack, Callee);
-        if (pResult)
+        if (pResult) {
+          m_functionsCheckedForRecursion[D] = pResult;
           return pResult;
+        }
       }
     }
     CallStack.erase(D);
+    m_functionsCheckedForRecursion[D] = nullptr;
     return nullptr;
   }
```